WordPress/Classes/ViewRelated/Post/PrivateSiteURLProtocol.m

```diff
@@ -114,51 +121,114 @@ - (void)startLoading
 {
     NSMutableURLRequest *mRequest = [self.request mutableCopy];
     [mRequest addValue:[NSString stringWithFormat:@"Bearer %@", [[self class] bearerToken]] forHTTPHeaderField:@"Authorization"];
-    NSURLSessionConfiguration *sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
-    self.session = [NSURLSession sessionWithConfiguration:sessionConfiguration delegate:self delegateQueue:nil];
-    self.sessionTask = [self.session dataTaskWithRequest:mRequest];
-    [self.sessionTask resume];
+    self.sessionTask = [[PrivateSiteURLProtocolSession sharedInstance] createSessionTaskForRequest:mRequest forProtocol:self];
+
 }
 
 - (void)stopLoading
 {
-    [self.sessionTask cancel];
-    [self.session invalidateAndCancel];
+    [[PrivateSiteURLProtocolSession sharedInstance] stopSessionTask:self.sessionTask];
     self.sessionTask = nil;
-    self.session = nil;
+}
+
+@end
+
+@interface PrivateSiteURLProtocolSession()
+
+@property (nonatomic, strong) NSMutableDictionary *taskToProtocolMapping;
+@property (nonatomic, strong) NSURLSession *session;
+
+@end
+
+@implementation PrivateSiteURLProtocolSession
+
++ (instancetype) sharedInstance
+{
+    static id _sharedInstance = nil;
+    static dispatch_once_t _onceToken;
+    dispatch_once(&_onceToken, ^{
+        _sharedInstance = [[self alloc] init];
+    });
+
+    return _sharedInstance;
+}
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        _taskToProtocolMapping = [NSMutableDictionary dictionary];
+    }
+    return self;
+}
+
+- (NSURLSession *)session
+{
+    if (_session == nil) {
+        NSURLSessionConfiguration *sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
+        _session = [NSURLSession sessionWithConfiguration:sessionConfiguration delegate:self delegateQueue:nil];
```

Comment:
Question: when passing `nil` to to `delegateQueue` the docs say a serial queue is used.  Would it be better to either create a new `NSOperationQueue` or perhaps edit the one that's created and set `maxConcurrentOperationCount` to `NSOperationQueueDefaultMaxConcurrentOperationCount`?